File: 021 Merge Two Sorted Lists.c
#include <stdio.h>
#include <stdlib.h>
/*
21 Merge Two Sorted Lists
https://oj.leetcode.com/problems/merge-two-sorted-lists/
*/
struct LinkedListNode{
    int rank;
    struct LinkedListNode* next;
};
typedef struct LinkedListNode node;

void printList(node* p){
    while(p != 0){
        printf("%d ",p->rank);
        p = p->next;
    }
    printf("\n");
}

node* mergeTwoSorted(node* la, node* lb){
    node *ret, *p;
    if(la == 0) return lb;
    if(lb == 0) return la;

    if(la->rank < lb->rank){
        ret = la;
        la = la->next;
    } else {
        ret = lb;
        lb = lb->next;
    }

    p = ret;
    while(la != 0 && lb != 0){
        if(la->rank < lb->rank){
            p->next = la;
            la = la->next;
        } else {
            p->next = lb;
            lb = lb->next;
        }
        p = p->next;
    }
    
    // notice here is different from "Merge 2 sorted array"
    if(la != 0) p->next = la;
    if(lb != 0) p->next = lb;
    return ret;
}

int main(){
    int repeat = 7; // will create repeat + 1 nodes
    int K;
    int i;
    node* head = (node*)malloc(sizeof(node));
    node* hea2 = (node*)malloc(sizeof(node));
    node* p;

    // init
    p = head;
    p->rank = 0;
    for(i=0;i<repeat;i++){
        p->next = (node*)malloc(sizeof(node));
        p = p->next;
        p->rank = i+1;
        p->next = 0;
    }
    printList(head);

    // init
    p = hea2;
    p->rank = 0;
    for(i=0;i<repeat;i++){
        p->next = (node*)malloc(sizeof(node));
        p = p->next;
        p->rank = i*2+1;
        p->next = 0;
    }
    printList(hea2);

    // test
    p = mergeTwoSorted(head,hea2);
    printList(p);

    //free
    head = p;
    while(head != 0){
        p = head->next;
        free(head);
        head = p;
    }
    return 0; 
}

File: 021 Merge Two Sorted Lists.java
/*
21. Merge Two Sorted Lists
https://leetcode.com/problems/merge-two-sorted-lists/

Merge two sorted linked lists and return it as a new sorted list.
The new list should be made by splicing together the nodes of the first two lists.

Example:

Input: 1->2->4, 1->3->4
Output: 1->1->2->3->4->4
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        // Use a fake head to make code shorter
        ListNode fakeHead = new ListNode();
        ListNode writer = fakeHead;
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val){
                writer.next = l1;
                l1 = l1.next;
            } else {
                writer.next = l2;
                l2 = l2.next;
            }
            writer = writer.next;
        }

        // Attach remaining tail
        if (l1 != null) {
            writer.next = l1;
        }
        if (l2 != null) {
            writer.next = l2;
        }
        
        ListNode res = fakeHead.next;
        fakeHead.next = null;
        return res;
    }
}
